Donor Page View API + Frontend Add Donor + Donate Functionality

ysiddhapura opened the pull request:
Donor Feature work in progress

Refer to https://github.com/ChicoState/PantryNode/issues/193

parthpandey1 commented:
Created a GET endpoint to fetch all users with at least one donation. 
Steps to test:
1. Pull from 'donorlookup' branch.
2. docker-compose up
3. curl http://localhost:3001/donors

Should return a JSON with person_id, full name and email of all donors.
![Screenshot 2023-05-01 011112](https://user-images.githubusercontent.com/69731623/235427367-a4b1e57f-c14e-43ce-84d2-460c6aa3bf6f.jpg)

briswells commented:
@parthpandey1 Code look awesome, and works correctly. Could you add authentication to the route. Just passing ensureAuthenticated as a route argument does not cause it to require authentication. You can look at the /feed route for an example.

briswells commented:
Looks great. Sorry I didn't notice this earlier but I would suggest is changing the names of the query parameters to be fname, and lname, just to keep everything consistent. You could also map the return dict so it returns data using the same dict keys that are being used for the API query parameter. Although this may cause confusion as different user routes might return the same fields named different things.

parthpandey1 commented:
@ysiddhapura @briswells @Anoushka444 @jayrevolinskyjr 
Added `/donations/:person_id` endpoint as part of https://github.com/ChicoState/PantryNode/pull/217/commits/240afa84fffec2885a34ce71542c4d0e871103fd.
This endpoint fetches donations made by a donor. 

Steps to test:

1. Setup postman as mentioned in https://github.com/ChicoState/PantryNode/pull/217#issuecomment-1531149852
2. Pull from 'donorlookup' branch.
3. docker-compose up.
4. Send GET request in this format - http://localhost:3001/donations/<person_id> . For example: `http://localhost:3001/donations/2`
5. This should fetch all the transactions with trans_type 'donation' made by the person. The JSON response includes donor details, transaction details and items within a transaction. Sample screenshot:
![Screenshot 2023-05-03 143702](https://user-images.githubusercontent.com/69731623/236058496-8e0fe3cd-2ed3-4a9e-9082-22a6f1e78b22.jpg)

This closes the 5th BE task in https://github.com/ChicoState/PantryNode/issues/193

Also fixed query parameters in lookupDonor endpoint as requested by @briswells as part of this commit. Now the request url should be formatted like http://localhost:3001/lookupDonor?fname=John&lname=Doe or http://localhost:3001/lookupDonor?person_id=1 or http://localhost:3001/lookupDonor?fname=John&lname=Doe&person_id=1.

parthpandey1 commented:
@ysiddhapura @briswells @Anoushka444 @jayrevolinskyjr 
Added a new POST endpoint as part of https://github.com/ChicoState/PantryNode/pull/217/commits/3f14d1194311b02e944270aef5cca82c92bf22d0 which allows users to donate items.

To test:

1. Setup postman as mentioned in https://github.com/ChicoState/PantryNode/pull/217#issuecomment-1531149852.
2. Pull from 'donorlookup'.
3. docker-compose up.
4. Send a POST request with request url http://localhost:3001/donate 
5. Request body should be structured as
`{
  "person_id": 1,
  "siteDetails": {
    "name": "Paradise"
  },
  "items": [
    {
      "transItemsDetails": {
        "quantity": 5,
        "expiration": "2023-12-01"
      },
      "itemDetails": {
        "name": "Ketchup",
        "size": 10,
        "category": "Condiments"
      },
      "storageTypeDetails": {
        "stor_type": "Fridge"
      }
    },
    {
      "transItemsDetails": {
        "quantity": 3,
        "expiration": "2023-11-01"
      },
      "itemDetails": {
        "name": "Lays",
        "size": 8,
        "category": "Snacks"
      },
      "storageTypeDetails": {
        "stor_type": "Dry-Storage"
      }
    }
  ]
}`

6. The response should return a JSON containing details of each record inserted. For example: 
`{
    "Site": {
        "site_id": 2,
        "name": "Paradise",
        "addr_id": 1
    },
    "Transaction": {
        "trans_id": 3,
        "person_id": 1,
        "date": "2023-05-03T22:28:32.710Z",
        "trans_type": "donation",
        "site": 2
    },
    "Items": [
        {
            "StorageType": {
                "stor_id": 4,
                "stor_type": "Fridge"
            },
            "Item": {
                "item_id": 1103,
                "name": "Ketchup",
                "size": 10,
                "category": "Condiments",
                "stor_id": 4,
                "barcode": null
            },
            "TransItem": {
                "trans_item_id": 6,
                "quantity": 5,
                "expiration": "2023-12-01",
                "trans_id": 3,
                "item_id": 1103
            }
        },
        {
            "StorageType": {
                "stor_id": 5,
                "stor_type": "Dry-Storage"
            },
            "Item": {
                "item_id": 1104,
                "name": "Lays",
                "size": 8,
                "category": "Snacks",
                "stor_id": 5,
                "barcode": null
            },
            "TransItem": {
                "trans_item_id": 7,
                "quantity": 3,
                "expiration": "2023-11-01",
                "trans_id": 3,
                "item_id": 1104
            }
        }
    ]
}`
7. To test whether the items are inserted correctly, /donations/:person_id endpoint can be used.
8. The transaction date is handled in the backend so no need to send it in the request body from frontend.
9. It is assumed that, site name, item quantity, expiration date, item name, item size, category and storage type will be manually entered by the user on FE to generate request body.

Screenshot from Postman:
![Screenshot 2023-05-03 154724](https://user-images.githubusercontent.com/69731623/236067192-6d0fc17a-ace5-4b66-b4fe-0b6b447b0f95.jpg)

This completes the 3rd BE task in https://github.com/ChicoState/PantryNode/issues/193.

ysiddhapura commented:
Have committed the below tasks: 

-Donor Lookup Page populated by Function/Query that returns all Donor ID, Names, Emails
(Not sorting properly currently), Successfully connected to /donors endpoint, to populate page with current donors.
-'Add Donor' Button linked to NewDonor(Name) and 'NewDonor(Name)' use Sequelize to create a new user in the DB table
Successfully created an endpoint, /addDonor to backend, using POST method, connected it to frontend, successfully adding donors, but can only see newly added ones in current instance, when you re-login, the new donors don't show up on the frontend.
- 'LookupDonor(Name/Id)' has checkbox to flip a Bool for setting Anonymous donor with hardcoded values
Created a form when pressing Lookup Donor, and added a frontend bool value isAnonymous, to mark anonymous donors and autofill the form.
-Also changed donors_test.tsx - to update tests with the new donor page functionality

jayrevolinskyjr commented:
I've opened this PR for review since it has so much code and checks off 13 out of the 16 [original subtasks](https://github.com/ChicoState/PantryNode/issues/193#issue-1682051340). The remaining tasks are broken down into #232, #233, #234. 

**Heads-up to any code reviewers**, there was a significant amount of UI edits merged to main that generated conflicts, resolving those conflicts generated code bloat in this branch.

jayrevolinskyjr reviewed (approved):
Runs like a charm on my machine
